src/compiler/binder.ts

```diff
@@ -1339,6 +1366,45 @@ namespace ts {
             bindExportAssignment(node);
         }
 
+        function bindThisPropertyAssignment(node: BinaryExpression) {
+            if (container.kind === SyntaxKind.FunctionExpression || container.kind === SyntaxKind.FunctionDeclaration) {
+                container.symbol.members = container.symbol.members || {};
+                declareClassMember(node, SymbolFlags.Property, SymbolFlags.PropertyExcludes);
+            }
+        }
+
+        function bindPrototypePropertyAssignment(node: BinaryExpression) {
+            // We saw a node of the form 'x.prototype.y = z'.
+            // This does two things: turns 'x' into a constructor function, and
+            // adds a member 'y' to the result of that constructor function
+            // Get 'x', the class
+            const classId = <Identifier>(<PropertyAccessExpression>(<PropertyAccessExpression>node.left).expression).expression;
+
+            // Look up the function in the local scope, since prototype assignments should immediately
+            // follow the function declaration
+            const funcSymbol = container.locals[classId.text];

```

Comment:
This doesn't guarantee that the prototype assignment *immediately* follows. Also, what if `classId` denotes a non-function, such as a variable or an interface declaration? I'm thinking you should instead check that the prototype assignment is immediately contained in a statement list and that the immediately preceding statement is a function declaration.